src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs

```diff
@@ -314,6 +403,194 @@ private void ResolveAsymmetricAlgorithm(AsymmetricSecurityKey key, string algori
+                {
+                    if (webKey.D == null)
+                        throw LogHelper.LogArgumentNullException(nameof(webKey.D));
+
+                    byte[] d = Base64UrlEncoder.DecodeBytes(webKey.D);
+                    foreach (byte b in d)
+                        Marshal.WriteByte(keyBlobPtr, index++, b);
+
+                    Marshal.Copy(keyBlobPtr, keyBlob, 0, keyBlob.Length);
+                    using (CngKey cngKey = CngKey.Import(keyBlob, CngKeyBlobFormat.EccPrivateBlob))
+                    {
+                        _ecdsa = new ECDsaCng(cngKey);
+                    }
+                }
+                else
+                {
+                    Marshal.Copy(keyBlobPtr, keyBlob, 0, keyBlob.Length);
+                    using (CngKey cngKey = CngKey.Import(keyBlob, CngKeyBlobFormat.EccPublicBlob))
+                    {
+                        _ecdsa = new ECDsaCng(cngKey);
+                    }
+                }
+                keyBlobHandle.Free();
+            }
+            finally
+            {
+                if (keyBlobHandle != null)
+                    keyBlobHandle.Free();
+            }
+        }
+
+        /// <summary>
+        /// Returns the size of key in bytes
+        /// </summary>
+        /// <param name="curveId">represents ecdsa curve -P256, P384, P512</param>
+        /// <returns>size of the key in bytes</returns>
+        private uint GetKeyByteCount(string curveId)
+        {
+            if (string.IsNullOrEmpty(curveId))
+                throw LogHelper.LogArgumentNullException(nameof(curveId));
+
+            uint keyByteCount;
+            switch (curveId)
+            {
+                case JsonWebKeyECTypes.P256:
+                    keyByteCount = 32;
+                    break;
+                case JsonWebKeyECTypes.P384:
+                    keyByteCount = 48;
+                    break;
+                case JsonWebKeyECTypes.P512:
+                    keyByteCount = 64;
+                    break;
+                default:
+                    throw LogHelper.LogException<ArgumentException>(LogMessages.IDX10645, curveId);
+            }
+            return keyByteCount;
+        }
+
+        /// <summary>
+        /// Returns the magic value representing the curve corresponding to the curve id.
+        /// </summary>
+        /// <param name="curveId">represents ecdsa curve -P256, P384, P512</param>
+        /// <param name="willCreateSignatures">whether the provider will create signatures or not</param>
+        /// <returns>uint representing the magic number</returns>
+        private uint GetMagicValue(string curveId, bool willCreateSignatures)
+        {
+            if (string.IsNullOrEmpty(curveId))
+                throw LogHelper.LogArgumentNullException(nameof(curveId));
+
+            KeyBlobMagicNumber magicNumber;
+            switch (curveId)
+            {
+                case JsonWebKeyECTypes.P256:
+                    if (willCreateSignatures)
```

Comment:
If enum is uint, then you can do away with local variable (magicNumber) and just return.